### ISSUE TITLE
No data in `CfdState`

### ISSUE THREAD

**Issue by da-kami** [itchysats/itchysats#290]
Motivation:

The `CfdState` is unnecessarily complex. We have to pass information on from state to state. That is unnecessary and annoying when matching. It bloats everything up and, on top of that, saving in the database is unnecessarily complicated. (Gigantic json).

Proposed refactor:

1. Move all data (dlc, attestation, cet-expiry, ...) onto the `Cfd`, the `CfdState` should not know data. (note: will require to extend the database to handle those fields - can be done via nullable fields or extension tables)
2. Remove `CfdStateCommon` and let the `transition_timestamp` be managed in the database.
3. Adapt the decision logic in `Cfd::handle` to reason about the data we know and the `CfdState` we are in.

**Comment by da-kami** [comment]
Any more thoughts @thomaseizinger ? 
I think once we have `collaborative settlement` in (and "freeze") we can tackle this one. It should simplify a lot of things.

**Comment by da-kami** [comment]
Sidenote: Our state transitions will be less "tyepsafe" once we do this change. At the moment each state ensures that information is present, by forcing the data onto the state. But I would still change this paradigm, because it makes things unnecessarily complicated.

**Comment by thomaseizinger** [comment]
> Any more thoughts @thomaseizinger ?

What I would add is that we should split up the `Cfd` model according to its life-cycle. IMO an instance of `Cfd` should only exist from when it is `Open`. Prior to it being open, we should use a different data structure that doesn't know about transaction IDs, CETs and all of that stuff.

>  the CfdState should not know data. (note: will require to extend the database to handle those fields - can be done via nullable fields or extension tables)

We might not need a dedicated enum for this then but instead do something like:

```rust
let is_committed = self.commit_tx.is_some();
```

**Comment by da-kami** [comment]
> > Any more thoughts @thomaseizinger ?
> 
> What I would add is that we should split up the `Cfd` model according to its life-cycle. IMO an instance of `Cfd` should only exist from when it is `Open`. Prior to it being open, we should use a different data structure that doesn't know about transaction IDs, CETs and all of that stuff.

I wonder if we should do this separate (on top) or before or alltogether - it feels like it is an orthogonal change. 
 
> > the CfdState should not know data. (note: will require to extend the database to handle those fields - can be done via nullable fields or extension tables)
> 
> We might not need a dedicated enum for this then but instead do something like:
> 
> ```rust
> let is_committed = self.commit_tx.is_some();
> ```

I agree that we *could* optimize this to not use an enum, but I would still keep the enum because it makes it easier to reason about what happened when reading data from the db, i.e. a list of state transitions that happened are easy to processed by a human. I had it several times that I tested the sanity by loading all states of a cfd to have a look what happen. Not sure that is a good argument though 😅

If you think that both changes should be tackled in some form when tackling the issue we should probably adapt the description :)

**Comment by thomaseizinger** [comment]
> I agree that we _could_ optimize this to not use an enum, but I would still keep the enum because it makes it easier to reason about what happened when reading data from the db, i.e. a list of state transitions that happened are easy to processed by a human.

What you want to avoid is redundant data. For example:

What is the point of having an enum that says `OpenCommitted` but `commit_tx` is `None`?

**Comment by da-kami** [comment]
> > I agree that we _could_ optimize this to not use an enum, but I would still keep the enum because it makes it easier to reason about what happened when reading data from the db, i.e. a list of state transitions that happened are easy to processed by a human.
> 
> What you want to avoid is redundant data. For example:
> 
> What is the point of having an enum that says `OpenCommitted` but `commit_tx` is `None`?

Let me rephrase my thinking: I think it would be nice to have some kind of list of "events" / "states" (...) that makes it easy for a human to reason about "what happened to a certain cfd" in case something goes wrong. At the moment I can load all the states for a cfd and aI basically see what happened. With only having data on the cfd that will become harder, but we can potentially record something else (e.g. state change events, ...) separately from the cfd that make it easy to reason about what happened.

**Comment by DeliciousHair** [comment]
Less on the internal model but more on the DB side of things, I think the data model can be trimmed down quite a bit via something like:
```sql
create table if not exists orders
(
    id                  integer primary key autoincrement,
    uuid                text unique not null,
    trading_pair        text        not null,
    position            text        not null,
    initial_price       text        not null,
    min_quantity        text        not null,
    max_quantity        text        not null,
    leverage            integer     not null,
    liquidation_price   text        not null,
    creation_timestamp  integer     not null,
    term                integer     not null,
    origin              text        not null,
    oracle_event_id     text        not null
);

create table if not exists cfds
(
    id           integer primary key autoincrement,
    order_id     integer unique not null,
    quantity_usd text           not null,

    foreign key (order_id) references orders (id)
);

create table if not exists state_tags (
    tag     integer primary key,
    value   text    not null
);

insert into state_tags (tag, value) values (
    -- all states as key-value pairs
);

create table if not exists cfd_states
(
    id                          integer     primary key autoincrement,
    cfd_id                      integer     not null,
    state_tag                   integer     not null,
    state_transition_timesamp   integer     not null
    foreign key (cfd_id) references cfds (id)
);
```
The idea being that the `cfd_states` table will now only contain some integer values and not take up as much space as the text fields--which only have a fixed number of possible values--that are moved to a lookup table instead. This could be extended to a bunch of the `orders` fields as well, but it's probably not as important given that that table grows much more slowly than the `cfd_states` table.

**Comment by DeliciousHair** [comment]
Partially incorporating into #357

**Comment by da-kami** [comment]
> Less on the internal model but more on the DB side of things, I think the data model can be trimmed down quite a bit via something like:
> 
> ```sql
> create table if not exists orders
> (
>     id                  integer primary key autoincrement,
>     uuid                text unique not null,
>     trading_pair        text        not null,
>     position            text        not null,
>     initial_price       text        not null,
>     min_quantity        text        not null,
>     max_quantity        text        not null,
>     leverage            integer     not null,
>     liquidation_price   text        not null,
>     creation_timestamp  integer     not null,
>     term                integer     not null,
>     origin              text        not null,
>     oracle_event_id     text        not null
> );
> 
> create table if not exists cfds
> (
>     id           integer primary key autoincrement,
>     order_id     integer unique not null,
>     quantity_usd text           not null,
> 
>     foreign key (order_id) references orders (id)
> );
> 
> create table if not exists state_tags (
>     tag     integer primary key,
>     value   text    not null
> );
> 
> insert into state_tags (tag, value) values (
>     -- all states as key-value pairs
> );
> 
> create table if not exists cfd_states
> (
>     id                          integer     primary key autoincrement,
>     cfd_id                      integer     not null,
>     state_tag                   integer     not null,
>     state_transition_timesamp   integer     not null
>     foreign key (cfd_id) references cfds (id)
> );
> ```
> 
> The idea being that the `cfd_states` table will now only contain some integer values and not take up as much space as the text fields--which only have a fixed number of possible values--that are moved to a lookup table instead. This could be extended to a bunch of the `orders` fields as well, but it's probably not as important given that that table grows much more slowly than the `cfd_states` table.

Yes! That is what we want, a properly modeled, normalized `n:m` relationship between `cfds` and "state data" (which is basically all data that can be added to the `Cfd`. 
If we can come up with a better name for `state_tags` (representing all data that can be added to the cfd once we learn it) that might help understanding. I cannot come up with a better name at the moment though.

**Comment by thomaseizinger** [comment]
Closing because outdated, need to re-write with latest understanding.
